### PR TITLE
Expose Pipeline primitives to end user

### LINKED ISSUE #502: App developers should be able to override the pipeline and supply their own context
Currently, the pipeline policies are controlled by the SDK author, and the end-user (app developer) can only customize them in ways exposed by the SDK through initializers or other convenience methods. As an escape hatch, we should allow the end user to create, modify and supply their own structure of pipeline policies, including custom policies, and since policies are stateless, they must also be able to submit their own pipeline context objects.

### LINKED ISSUE #509: Storage and Chat Library client initializers do not follow guidelines
https://azure.github.io/azure-sdk/ios_design.html#client-interface

The initializer here differs from that used in AzureStorageBlob and that generated by Autorest.Swift

**Guidelines**
```public init(endpoint: String, credential: String? = nil, options: Options? = nil)```

**Actual**
```public init(baseUrl: URL, authPolicy: Authenticating, withOptions options: AzureOptions? = nil)```

The "right" answer probably lies closer to the guidelines.

### PR DESCRIPTION
**Pull request by tjprescott**
Fixes #509 
Fixes #502

### CODE REVIEW

File: sdk/core/AzureCore/Source/Pipeline/Policies/ContentDecodePolicy.swift [hunk 1]
```diff
@@ -194,8 +194,8 @@ public class ContentDecodePolicy: PipelineStage {
         else { return }
         contentType = contentType.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
         do {
-            if let deserializedData = try deserialize(from: returnResponse, contentType: contentType) {
-                returnResponse.add(value: deserializedData as AnyObject, forKey: .deserializedData)
+            if let deserializedData = try deserialize(from: returnResponse, contentType: contentType) as? AnyObject {
```
Comment (bsiegel) [inline sdk/core/AzureCore/Source/Pipeline/Policies/ContentDecodePolicy.swift:197]: Unsure why this changed - does this affect behavior or is it just cosmetic?

File: sdk/communication/AzureCommunicationChat/Source/AzureCommunicationChatClient.swift [hunk 2]
```diff
@@ -109,7 +109,7 @@ public final class AzureCommunicationChatClient: PipelineClient, PageableClient
         }
 
         // Send request
-        let context = PipelineContext.of(keyValues: [
+        let context = options?.context ?? PipelineContext.of(keyValues: [
```
Comment (bsiegel) [inline sdk/communication/AzureCommunicationChat/Source/AzureCommunicationChatClient.swift:112]: Should we be merging in the "default" context entries here rather than just taking the user's context object as-is? I can see arguments both ways - the user shouldn't need to know to set the allowed status codes for example, but also if for some reason the service changes and a new status code is added the user needs an escape hatch to work around our default value for that context key.

File: sdk/communication/AzureCommunicationChat/Source/Options/AzureCommunicationChatClientOptions.swift [hunk 3]
```diff
@@ -26,6 +26,11 @@ public struct AzureCommunicationChatClientOptions: AzureClientOptions {
     public let transportOptions: TransportOptions
     /// The default dispatch queue on which to call all completion handler. Defaults to `DispatchQueue.main`.
     public let dispatchQueue: DispatchQueue?
+    /// An array of `PipelineStage` policies to use in lieu of the default ones.
+    public let pipeline: [PipelineStage]?
```
Comment (bsiegel) [inline sdk/communication/AzureCommunicationChat/Source/Options/AzureCommunicationChatClientOptions.swift:30]: Unclear whether this is what the arch board intended, or whether they were asking for some way to inject additional policies at the end of the pipeline instead. I thought I had noted that it was the latter but I could be wrong. I think that might make sense regardless since a policy at the end of the pipeline could pretty much overwrite anything done by any earlier policy.

File: sdk/communication/AzureCommunicationChat/Source/Options/AzureCommunicationChatClientOptions.swift [hunk 4]
```diff
@@ -26,6 +26,11 @@ public struct AzureCommunicationChatClientOptions: AzureClientOptions {
...
+    /// An `HTTPTransportStage` policy to use in lieu of the default.
+    public let transport: HTTPTransportStage?
```
Comment (bsiegel) [inline sdk/communication/AzureCommunicationChat/Source/Options/AzureCommunicationChatClientOptions.swift:32]: Can we add these (pipeline, transport) to `TransportOptions` instead? That way they don't have to be repeated in every ClientOptions object and they seem like they make more sense there anyhow.

**Review by bsiegel — CHANGES_REQUESTED** [review]

File: sdk/core/AzureCore/Source/Pipeline/Policies/ContentDecodePolicy.swift (hunk 1 above)
Comment (tjprescott) [reply sdk/core/AzureCore/Source/Pipeline/Policies/ContentDecodePolicy.swift:197]: It was necessary with a variant I was trying, but it is no longer necessary so I removed it.

File: sdk/communication/AzureCommunicationChat/Source/Options/AzureCommunicationChatClientOptions.swift (hunk 4 above)
Comment (tjprescott) [reply sdk/communication/AzureCommunicationChat/Source/Options/AzureCommunicationChatClientOptions.swift:32]: Done.

File: sdk/communication/AzureCommunicationChat/Source/AzureCommunicationChatClient.swift (hunk 2 above)
Comment (tjprescott) [reply sdk/communication/AzureCommunicationChat/Source/AzureCommunicationChatClient.swift:112]: If we update the default context with any user keys, they should be able to do both.

File: sdk/communication/AzureCommunicationChat/Source/Options/AzureCommunicationChatClientOptions.swift (hunk 3 above)
Comment (tjprescott) [reply sdk/communication/AzureCommunicationChat/Source/Options/AzureCommunicationChatClientOptions.swift]: This also makes sense within the context of distinguishing between "per request" and "per retry" policies.

File: sdk/communication/AzureCommunicationChat/Source/AzureCommunicationChatClient.swift (hunk 2 above)
Comment (vcolin7) [reply sdk/communication/AzureCommunicationChat/Source/AzureCommunicationChatClient.swift:112]: @bsiegel Is it possible to offer both public APIs? (i.e. something like `replaceContext` and `mergeContext`).

**Comment by tjprescott** [comment]
@bsiegel pushed another iteration with an extensible pipeline. Please take a second look. I still need to add unit tests for extending the PipelineContext

**Comment by tjprescott** [comment]
@bsiegel I've added tests for the Pipeline and Context cases. This PR will not actually implement the RetryPolicy, but does include the logic for structuring the pipeline around the Retry policy.

File: sdk/communication/AzureCommunicationChat/Source/AzureCommunicationChatClient.swift (hunk 2 above)
Comment (bsiegel) [reply sdk/communication/AzureCommunicationChat/Source/AzureCommunicationChatClient.swift]: Sounds like the answer is that we just need essentially, `mergeContext`

File: sdk/storage/AzureStorageBlob/Source/Options/OptionsModels.swift [hunk 5]
```diff
@@ -154,7 +154,7 @@ public struct CustomerProvidedEncryptionKey: Codable, Equatable {
     }
 
     /// Specifies the algorithm to use when encrypting data using the given key. Must be AES256.
-    public let algorithm = "AES256"
+    var algorithm = "AES256"
```
Comment (bsiegel) [inline sdk/storage/AzureStorageBlob/Source/Options/OptionsModels.swift:157]: Unrelated change?

File: sdk/storage/AzureStorageBlob/Source/StorageBlobClient.swift [hunk 6]
```diff
@@ -84,16 +84,16 @@ public final class StorageBlobClient: PipelineClient {
...
     ) throws {
         self.options = options
         super.init(
-            baseUrl: baseUrl,
-            transport: URLSessionTransport(),
-            policies: [
-                UserAgentPolicy(for: StorageBlobClient.self, telemetryOptions: self.options.telemetryOptions),
+            endpoint: endpoint,
+            transport: options.transportOptions.transport ?? URLSessionTransport(),
+            policies: options.transportOptions.perRequestPolicies ?? [
```
Comment (bsiegel) [inline sdk/storage/AzureStorageBlob/Source/StorageBlobClient.swift:95]: Now that we merge policies I think this needs to be reverted

File: sdk/storage/AzureStorageBlob/Source/Options/OptionsModels.swift (hunk 5 above)
Comment (tjprescott) [reply sdk/storage/AzureStorageBlob/Source/Options/OptionsModels.swift:157]: It was to get rid of an XCode warning.

File: sdk/storage/AzureStorageBlob/Source/StorageBlobClient.swift (hunk 6 above)
Comment (tjprescott) [reply sdk/storage/AzureStorageBlob/Source/StorageBlobClient.swift:95]: Yes, it does.

File: sdk/storage/AzureStorageBlob/Source/Options/OptionsModels.swift (hunk 5 above)
Comment (bsiegel) [reply sdk/storage/AzureStorageBlob/Source/Options/OptionsModels.swift:157]: Hmm, odd. Is it just because it's unused? If so I guess maybe just delete it?

**Review by bsiegel — APPROVED** [review]
